File: Android/PerimeterX_SDK_Demo/app/src/main/java/com/perimeterx/perimeterx_sdk_demo/PerimeterxManager.kt
package com.perimeterx.android_sdk_demo

import android.app.Application
import android.os.Handler
import android.os.Looper
import androidx.core.os.postDelayed
import com.perimeterx.mobile_sdk.PerimeterX
import com.perimeterx.mobile_sdk.PerimeterXDelegate
import com.perimeterx.mobile_sdk.main.PXPolicy
import com.perimeterx.mobile_sdk.main.PXPolicyUrlRequestInterceptionType
import com.perimeterx.mobile_sdk.main.PXStorageMethod

object PerimeterxManager: PerimeterXDelegate {

    // PerimeterXDelegate

    override fun perimeterxRequestBlockedHandler(url: String?, appId: String) {
        println("PerimeterX Request Blocked")
    }

    override fun perimeterxChallengeSolvedHandler(appId: String) {
        println("PerimeterX Challenge Solved")
    }

    override fun perimeterxChallengeCancelledHandler(appId: String) {
        println("PerimeterX Challenge Cancelled")
    }

    override fun perimeterxHeadersWereUpdated(headers: HashMap<String, String>, appId: String) {
        println("PerimeterX Headers Were Updated")
    }

    // properties

    const val pxAppId = "PXj9y4Q8Em"

    // PerimeterX

    fun start(application: Application) {
        println("SDK version: ${PerimeterX.sdkVersion()}")

        start(application, pxAppId)
        setCustomParameters()
    }

    private fun start(application: Application, appId: String) {
        // Create and configure the policy //
        val policy = PXPolicy()
        policy.storageMethod = PXStorageMethod.DATA_STORE
        policy.urlRequestInterceptionType = PXPolicyUrlRequestInterceptionType.INTERCEPT_AND_RETRY_REQUEST
        policy.doctorCheckEnabled = true

        // Start PerimeterX SDK with your AppID //
        PerimeterX.start(application, appId, this, policy)
    }

    private fun setCustomParameters() {
        val customParameters = HashMap<String, String>()
        customParameters["custom_param1"] = "hello"
        customParameters["custom_param2"] = "world"
        PerimeterX.setCustomParameters(customParameters, null)
    }
}

File: iOS/PerimeterX_SDK_Demo/PerimeterX_SDK_Demo/PerimeterX/PerimeterxManager.swift
//
//  PerimeterxManager.swift
//  PerimeterX_SDK_Demo
//
//  Created by PerimeterX.
//

import Foundation
import PerimeterX_SDK

class PerimeterxManager: NSObject, PerimeterXDelegate {
    
    // MARK: - PerimeterXDelegate
    
    func perimeterxDidRequestBlocked(url: URL?, appId: String) {
        print("PerimeterX Request Blocked")
    }
    
    func perimeterxDidChallengeSolved(forAppId appId: String) {
        print("PerimeterX Challenge Solved")
    }
    
    func perimeterxDidChallengeCancelled(forAppId appId: String) {
        print("PerimeterX Challenge Cancelled")
    }
    
    func perimeterxHeadersWereUpdated(headers: [String : String], forAppId appId: String) {
        print("PerimeterX Headers Were Updated")
    }
    
    // MARK: - singleton
    
    static let shared = PerimeterxManager()
    
    // MARK: - properties
    
    private let pxAppId = "PXj9y4Q8Em"
    let urlRequestInterceptionType: PXPolicyUrlRequestInterceptionType = .interceptWithDelayedResponse
    
    // MARK: - PerimeterX
    
    func start() {
        print("SDK version: \(PerimeterX.sdkVersion())")
        
        start(appId: pxAppId)
        setCustomParameters()
    }
    
    private func start(appId: String) {
        // Create and config the policy //
        let policy = PXPolicy()
        policy.urlRequestInterceptionType = urlRequestInterceptionType
        policy.doctorCheckEnabled = true
        policy.set(domains: ["sample-ios.pxchk.net"], forAppId: appId)
        
        // Start PerimeterX SDK with your AppID //
        do {
            try PerimeterX.start(appId: appId, delegate: self, policy: policy)
        }
        catch {
            print("error: \(error)")
        }
    }
    
    private func setCustomParameters() {
        var customParameters = [String: String]()
        customParameters["custom_param1"] = "hello"
        customParameters["custom_param2"] = "world"
        do {
            try PerimeterX.setCustomParameters(parameters: customParameters)
        }
        catch {
            print("error: \(error)")
        }
    }
}
